# Lab 3


In [ ]:
# Definición de las matrices como listas de listas
M = [[120, 5],
     [5, 115]]

Mp = [[200, 10],
      [10, 1]]

M, Mp
def det2x2(A):
    return A[0][0]*A[1][1] - A[0][1]*A[1][0]

def trace2x2(A):
    return A[0][0] + A[1][1]



In [3]:
# Cálculo manual de eigenvalores para una matriz 2x2
def eigenvalues_2x2(A):
    a = 1
    b = -(A[0][0] + A[1][1])
    c = det2x2(A)
    
    discriminant = b*b - 4*a*c
    sqrt_disc = discriminant ** 0.5
    
    lambda1 = (-b + sqrt_disc) / (2*a)
    lambda2 = (-b - sqrt_disc) / (2*a)
    
    return lambda1, lambda2


In [4]:
eig_M = eigenvalues_2x2(M)
eig_M
eig_Mp = eigenvalues_2x2(Mp)
eig_Mp


(200.50124999218758, 0.4987500078124043)

Como visto en clase:

R = det(M) − k · (trace(M))²


In [6]:
k = 0.04

det_M = det2x2(M)
trace_M = trace2x2(M)

R_M = det_M - k * (trace_M ** 2)

det_M, trace_M, R_M


(13775, 235, 11566.0)

In [7]:
det_Mp = det2x2(Mp)
trace_Mp = trace2x2(Mp)

R_Mp = det_Mp - k * (trace_Mp ** 2)

det_Mp, trace_Mp, R_Mp


(100, 201, -1516.04)

Para la matriz M se obtuvo:

det(M) = 13775  
trace(M) = 235  
R = 11566  

El valor de R es positivo y significativamente grande. Según el criterio de Harris,
cuando R > 0 y los eigenvalores son grandes y similares, el píxel corresponde a una
ESQUINA, ya que existen cambios de intensidad importantes en ambas direcciones.

### Clasificación según eigenvalores y Harris

→ Representa una ESQUINA si: 
- Eigenvalores grandes y similares
- Respuesta de Harris positiva


→ Representa un BORDE si:
- Un eigenvalor grande y uno pequeño
- Respuesta de Harris negativa

